### PR TITLE
[GOBBLIN-712] Add version strategy pickup for ConfigBasedDataset distcp workflow

### PR DESCRIPTION
**Pull request by yukuai518**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-712

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
  consider to compare the data file version instead of always using data file length or modification timestamp.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by yukuai518** [comment]
@ibuenros can you help review?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java [hunk 1]
```diff
@@ -177,12 +231,30 @@ public String datasetURN() {
...
-          && copyToFileMap.get(newPath).getModificationTime() > originFileStatus.getModificationTime()) {
-        log.debug("Copy from timestamp older than copy to timestamp, skipped copy {} for dataset with metadata {}",
-            originFileStatus.getPath(), this.rc.getMetaData());
-      } else {
+
+      boolean shouldCopy = true;
+      if (copyToFileMap.containsKey(newPath)) {
+        if (this.srcDataFileVersionStrategy.get() instanceof ModTimeDataFileVersionStrategy) {
+          // default version strategy should also compare data file length
+          if (copyToFileMap.get(newPath).getLen() == originFileStatus.getLen()
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java]: I think the file length should be a different comparison altogether, which can be enabled or disabled independently.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java [hunk 2]
```diff
@@ -177,12 +231,30 @@ public String datasetURN() {
...
+              && copyToFileMap.get(newPath).getModificationTime() > originFileStatus.getModificationTime()) {
+            log.debug("Copy from timestamp older than copy to timestamp, skipped copy {} for dataset with metadata {}",
+                originFileStatus.getPath(), this.rc.getMetaData());
+            shouldCopy = false;
+          }
+        } else {
+          // other version strategy should just compare the version number
+          Comparable srcVer = this.srcDataFileVersionStrategy.get().getVersion(originFileStatus.getPath());
+          Comparable dstVer = this.dstDataFileVersionStrategy.get().getVersion(copyToFileMap.get(newPath).getPath());
+          if (srcVer.compareTo(dstVer) == 0) {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java]: You want to copy only if `srcVersion > dstVersion`, right?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java (hunk 1 above)
Comment (yukuai518) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java]: @ibuenros Let's assume you have the same file copied from source to destination. Now what if you apply different purger algorithm on both src and dst file? I think then the file size would be different, but you don't want to copy again I guess? So should our Purger version strategy exclude the length check?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java (hunk 2 above)
Comment (yukuai518) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java]: I think so, let me change it.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java (hunk 1 above)
Comment (yukuai518) [reply gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java]: Change a little bit.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java [hunk 3]
```diff
@@ -94,6 +107,40 @@ public ConfigBasedDataset(ReplicationConfiguration rc, Properties props, CopyRou
...
+      Config versionStrategyConfig = ConfigFactory.parseMap(ImmutableMap.of(
+          DataFileVersionStrategy.DATA_FILE_VERSION_STRATEGY_KEY, versionStrategy.isPresent()? versionStrategy.get() :
+              props.getProperty(DataFileVersionStrategy.DATA_FILE_VERSION_STRATEGY_KEY,
+                  ModTimeDataFileVersionStrategy.Factory.class.getName())));
+
+      DataFileVersionStrategy strategy = DataFileVersionStrategy.instantiateDataFileVersionStrategy(fs, versionStrategyConfig);
+      log.debug("{} has version strategy {}", hEndpoint.getClusterName());
+      return Optional.of(strategy);
+    } catch (IOException e) {
+      return Optional.absent();
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java]: Should we log the exception?

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java [hunk 4]
```diff
@@ -128,6 +175,17 @@ public String datasetURN() {
...
     }
 
+    if (!this.srcDataFileVersionStrategy.isPresent() || !this.dstDataFileVersionStrategy.isPresent()) {
+      log.warn("Version strategy doesn't exist, cannot handle copy");
+      return copyableFiles;
+    }
+
+    if (!this.srcDataFileVersionStrategy.get().getClass().getName()
+        .equals(this.dstDataFileVersionStrategy.get().getClass().getName())) {
+      log.warn("Version strategy doesn't match, cannot handle copy");
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java]: Let's print out the src and target strategies for easier debuggability.

File: gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java [hunk 5]
```diff
@@ -135,6 +135,7 @@ public ConfigBasedDataset(ReplicationConfiguration rc, Properties props, CopyRou
       log.debug("{} has version strategy {}", hEndpoint.getClusterName());
       return Optional.of(strategy);
     } catch (IOException e) {
+      log.error("Version strategy cannot be created due to {}", e.toString());
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/replication/ConfigBasedDataset.java]: Please use `log.error("Version strategy ...", e)`

**Comment by ibuenros** [comment]
+1

**Pull request closed without merging**